File: rt/src/ray.rs
use crate::vec3::{Point3, Vec3};

pub struct Ray {
    pub origin: Point3,
    pub direction: Vec3,
}

impl Ray {
    pub fn new(origin: &Point3, direction: &Vec3) -> Self {
        Self {
            origin: origin.clone(),
            direction: direction.clone(),
        }
    }

    pub fn at(&self, t: f64) -> Point3 {
        self.origin + t * self.direction
    }
}

File: in_C++/src/Ray.cpp
#include "Ray.hpp"

Ray::Ray(const Vec3& origin, const Vec3& direction)
    : m_origin(origin), m_direction(direction) {}

Ray& Ray::operator=(const Ray& r) {
    if (this == &r)
        return *this;

    if (r.direction().magnitude() == 0.0) {
        throw NullDirectionException();
    }

    this->m_origin = r.m_origin;
    this->m_direction = r.m_direction;
    return *this;
}

Point3 Ray::origin() const {
    return this->m_origin;
}

Vec3 Ray::direction() const {
    return this->m_direction;
}

Point3 Ray::at(const double t) const {
    return this->origin() + t * this->direction();
}
